network: introduce multipath routing

yuwata opened the pull request:
Closes #12541.
Replaces #14063.

cc @ssahani and @n0rad.

ssahani commented:
Thanks for working on it. Let's review this after 244. I am working on openvswitch integration.

yuwata commented:
@ssahani and @keszybz PTAL.

ssahani commented:
Look almost good . CI is failing.

```
======================================================================
FAIL: test_route_static (__main__.NetworkdNetworkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 1746, in test_route_static
    self.assertRegex(output, '2001:1234:5:7fff:ff:ff:ff:ff proto static metric 1024')
AssertionError: Regex didn't match: '2001:1234:5:7fff:ff:ff:ff:ff proto static metric 1024' not found in '2001:1234:5:7fff:ff:ff:ff:ff via 2001:1234:5:8fff:ff:ff:ff:ff dev dummy98 proto static metric 1024 pref medium\n2001:1234:5:7fff:ff:ff:ff:ff via 2001:1234:5:9fff:ff:ff:ff:ff dev dummy98 proto static metric 1024 pref medium'

----------------------------------------------------------------------
Ran 111 tests in 2201.561s

FAILED (failures=1, expected failures=13)
[TASK END] Fri  6 Dec 17:09:28 CET 2019
[RESULT] systemd-networkd-tests.py - FAIL (EC: 1) (log file: /root/testsuite-logs-upstream.rnA/systemd-networkd-tests.py_FAIL.log)

[TASK] coredumpctl_collect
Waiting for PID 567553 to finish

```

ssahani commented on man/systemd.network.xml:
```diff
@@ -1377,6 +1377,16 @@
+          <term><varname>MultiPathRoute=<replaceable>address</replaceable> [dev=<replaceable>name</replaceable>] [weight=<replaceable>number</replaceable>]</varname></term>
+          <listitem>
+            <para>Configures multipath route. Takes gateway address. Optionally, takes a network
+            interface name or index prefixed with <literal>dev=</literal>, and a weight in 1..256
```
May be add `Multipath routing is the routing technique of using multiple alternative paths through a network.

ssahani commented on src/network/networkd-route.c:
```diff
@@ -516,6 +518,61 @@ int route_expire_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+}
+
+static int append_nexthops(Route *route, sd_netlink_message *req) {
+        char buf[4096] = {};
```
Reverse tree order looks good no ?

ssahani commented on src/network/networkd-route.c:
```diff
@@ -516,6 +518,61 @@ int route_expire_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+
+        ORDERED_SET_FOREACH(m, route->multipath_routes, i) {
+                r = append_nexthop_one(route, m, rta, 4096, rtnh);
+                if (r < 0)
```
May be add a debug log here.

yuwata replied:
The caller of this function logs about errors. So, I do not think it is necessary to add logs here.

yuwata commented:
@ssahani Thank you for the review. I addressed your comment except for the one about the debug log. And now I added a test for the case that IPv4 destination with IPv6 gateways (`RTA_VIA`). PTAL.

ssahani commented:
LGTM

poettering commented on man/systemd.network.xml:
```diff
@@ -1381,6 +1381,17 @@
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>MultiPathRoute=<replaceable>address</replaceable> [dev=<replaceable>name</replaceable>] [weight=<replaceable>number</replaceable>]</varname></term>
```
i think we should try to avoid "double keying" settings, i.e. have additional config keys inside of config key values. Can't we fine a different way to configure this? 

at the least, maybe we can do away with the key names, i.e. instead write it like this:

MultiPathRoute=address@iface weight

i.e. do something like we did for IO bandwidth settings in unit files, where you write:

IODeviceBandwidth=/dev/sda 5M

I mean, the simple ini file concept we generally use is very simplistic, and it isn't really set up to deal with complex objects, but I'd always take this as encouragement to find way to "flatten out" these objects psychologically and hide the fact that some things are multidimensional, by pretending they are uni-dimensional... Hence, using explicit "subfield" names make clear these are multidimensional, but if we instead just have pairs of space separated values this doesn't become as obvious, if you follow at all what I am babbling here...

poettering commented on src/network/networkd-route.c:
```diff
@@ -516,6 +518,65 @@ int route_expire_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+                .rtnh_len = sizeof(*rtnh),
+                .rtnh_ifindex = m->ifindex,
+                .rtnh_hops = m->weight > 0 ? m->weight - 1 : 0,
+        };
```
hmm, I am generally not a fan of functions that modify parameters on failure. Can't we find a nicer way for ths? i.e. isn't it enough to move these assignments to the end of this function when we know that everything worked correctly?

poettering commented on src/network/networkd-route.c:
```diff
@@ -516,6 +518,65 @@ int route_expire_handler(sd_event_source *s, uint64_t usec, void *userdata) {
+        struct rtnexthop *rtnh;
+        struct rtattr *rta;
+        MultipathRoute *m;
+        char buf[4096];
```
where does this size come from? shouldn't this grow automatically (via GREEDY_REALLOC for example)

yuwata commented:
@poettering Thank you for the review. I've force-pushed a revised version. Now I hope all your comments are addressed. PTAL.

mrc0mmand commented:
Almost there:

```
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]: ==30848==ERROR: AddressSanitizer: attempting free on address which was not malloc()-ed: 0x7ffd655758c0 in thread T0
Dec 09 22:01:03 arch.localdomain systemd-udevd[30850]: Using default interface naming scheme 'v243'.
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #0 0x7fe2680917c9 in free (/usr/lib/clang/9.0.0/lib/linux/libclang_rt.asan-x86_64.so+0xeb7c9)
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #1 0x55bc7f4b25a5 in freep /build/build/../src/basic/alloc-util.h:83:9
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #2 0x55bc7f4d00d9 in config_parse_multipath_route /build/build/../src/network/networkd-route.c:1676:1
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #3 0x7fe26722c889 in next_assignment /build/build/../src/shared/conf-parser.c:132:32
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #4 0x7fe267220b8f in parse_line /build/build/../src/shared/conf-parser.c:270:16
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #5 0x7fe26721df4e in config_parse /build/build/../src/shared/conf-parser.c:391:21
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #6 0x7fe26722109f in config_parse_many_files /build/build/../src/shared/conf-parser.c:448:21
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #7 0x7fe267221e7e in config_parse_many /build/build/../src/shared/conf-parser.c:507:16
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #8 0x55bc7f4881b6 in network_load_one /build/build/../src/network/networkd-network.c:461:13
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #9 0x55bc7f48976d in network_load /build/build/../src/network/networkd-network.c:536:21
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #10 0x55bc7f460574 in manager_load_config /build/build/../src/network/networkd-manager.c:1870:13
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #11 0x55bc7f430959 in run /build/build/../src/network/networkd.c:87:13
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #12 0x55bc7f42f04d in main /build/build/../src/network/networkd.c:130:1
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #13 0x7fe266a1b152 in __libc_start_main (/lib64/libc.so.6+0x27152)
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]:     #14 0x55bc7f42ef4d in _start (/build/build/systemd-networkd+0x255f4d)
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]: Address 0x7ffd655758c0 is located in stack of thread T0
Dec 09 22:01:03 arch.localdomain systemd-networkd[30848]: SUMMARY: AddressSanitizer: bad-free (/usr/lib/clang/9.0.0/lib/linux/libclang_rt.asan-x86_64.so+0xeb7c9) in free
```

Possibly relevant full coredump from CentOS 7: https://ci.centos.org/job/systemd-pr-build/9232/artifact//systemd-centos-ci/artifacts_I6h7a0/testsuite-logs-upstream.hSQ/coredumpctl_collect_FAIL.log

yuwata commented:
@mrc0mmand Thank you for catching the issue. It is now fixed. PTAL.

poettering commented on src/libsystemd/sd-netlink/netlink-util.c:
```diff
@@ -178,3 +179,60 @@ int rtnl_log_parse_error(int r) {
+
+void rtattr_append_attribute_internal(struct rtattr *rta, unsigned short type, const void *data, size_t data_length) {
+        size_t padding_length;
+        char *padding;
```
doesn't matter much, but to me "char" indicates this was actually character data... I'd always use "uint8_t" for this, since these are "arbitrary padding bytes"

poettering commented:
CI issue appears unrelated. Merging.

aep commented:
bit late maybe but its not entirely clear how this is supposed to work at all.
a  [Route] is attached to a network in systemd-networkd, so its removed if one link in an ecmp goes down.

That's not.. "multi" path?

the test also indicates this only works when both routes are on the same link, but whats the point of that

https://github.com/systemd/systemd/pull/14194/files#diff-61992e6b0355a76830b1a4fa191babb3532354a3f625edf802ff20acb1adf23cR82

yuwata commented:
@aep

> the test also indicates this only works when both routes are on the same link, but whats the point of that

The setting can take another interface name. Though, actually it is not tested.

> a [Route] is attached to a network in systemd-networkd, so its removed if one link in an ecmp goes down.

Could you open a new issue page with more details?
